### ISSUE TITLE
how to add extensible (platform specific) resources in a chart?

### ISSUE THREAD

**Issue by jstrachan** [kubernetes/helm#1331]
Where possible developers should create charts that run on any kubernetes cluster - portability of cloud native apps is one of the big value propositions of kubernetes!

However some kubernetes platforms have their own custom resources that an application developer may wish to use - but only if the app is running on that platform. e.g. OpenShift adds a few extension resources like OAuthClient, ImageStream, BuildConfig. [API Federation](https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/federated-api-servers.md) may become a common way to extend kubernetes with extra capabilities. 

This reminds me a little of JEE world (in a good way - honest! ;) where there were standard manifests that worked everywhere and different app server providers could have their own custom manifests to add extended metadata which were ignored by other providers but gave you finer grained configuration over behaviour on specific platforms.

So I just wanted to ponder how we might tackle this.

One option is every platform maintain their own fork of helm. (Maybe its just `Tiller` that needs to be forked actually which wouldn't be too bad maybe?). But then this means a chart must be either for vanilla kubernetes or for openshift or tectonic or GKE or whatever. I prefer the idea of shared chart repositories and each chart can run anywhere.

I guess the easiest option today is to use [ThirdPartyResources](https://github.com/kubernetes/kubernetes/blob/master/docs/design/extending-api.md) for all the extensions; and assume that there's some microservice to process it on each platform and if there isn't its just some metadata which may be useful anyways even if its not processed?

Or another approach is to support platform specific folders where we can put manifests which are only used on those platforms? 

e.g. maybe we use a canonical platform ID as a folder inside the charts? The following example is a little over the top mind you but just shows some possibilities:

```
wordpress/
  Chart.yaml
  templates/
    openshift.org/
      oauthclient.yaml
    tectonic.io/
     something.yaml
    google.com/
      googlePVs.yaml
```

Then on one of those platforms, say `openshift.org` we could register a microservice into tiller to handle applying resources in that folder? We could find these microservices using a label/annotation of, say, `tiller = openshift.org` then just post the resource to that microservices service endpoint?

Thoughts?

**Comment by technosophos** [comment]
This is blocked by #1295, which seems like it ought to be simple, but which adds some additional check in Kubernetes that is causing our tests to fail.

Other than that, the way Helm works is it dynamically queries the `versions` endpoint to learn about the kinds that the cluster supports. If the cluster says it supports them, Helm will try to push them into the cluster.

But I'm not sure what the best way is to make one chart support certain subsets of features depending on the cluster (short of using a `values.yaml` thing like `{{if .Values.clusterType == "openshift"}}`

So I'm really open to ideas on this one.

**Comment by jstrachan** [comment]
@technosophos ah thanks for the heads up! Yeah I'd like to avoid having to have lots of fragmentation in chart repositories and charts if we can. I can see various extensions coming along for different kubernetes cluster implementations; that a chart may wish to take advantage of if its present - but always work on any kubernetes cluster too.

I wonder if we could mark certain templates as being optional - so that if a cluster doesn't support them then its OK not to deploy them?
e.g.

```
wordpress/
  Chart.yaml
  templates/
    foo.yaml
    optional/
      bar.yaml
```
Then anything in the `optional` folder is considered optional - depending on if the cluster has those `versions` available? It could log that they are not available but would not fail if they could not be applied?

Then it'd be nice and simple to separate `mandatory` and `optional` templates; anything in the `optional` folder are best effort and ignored if there's no `versions` endpoint for those kinds?

**Comment by jstrachan** [comment]
Incidentally I just tried to use a chart with some openshift resources inside it. I then used helm inside openshift (using [minishift](https://github.com/jimmidyson/minishift) to boot up a cluster on my laptop). I got these results:
```
Error: release cranky-gorilla failed: [unable to decode "": no kind "DeploymentConfig" is registered for version "v1", unable to decode "": no kind "OAuthClient" is registered for version "v1"]
```
the kinds of the openshift resources look like this
```yaml
apiVersion: "v1"
kind: "DeploymentConfig"
```
I guess the issue is that really, these resources are not at `/api/v1` in the API server, in openshift they are at `/oapi/v1` and from the YAML there's no way for helm to know that really.

So its looking like, unless we put these inside ThirdPartyResources or something, we would need a tiller fork that was capable of detecting the OpenShift specific resources

**Comment by jstrachan** [comment]
here's a sample gogs template using OpenShift resources (DeploymentConfig and OAuthClient) in case anyone wants to play: 
[gogs.zip](https://github.com/kubernetes/helm/files/542633/gogs.zip)

**Comment by technosophos** [comment]
Yeah, we could do that, or we could allow a user to bypass the checks. However, I am assuming that you are also using a custom `kubectl` library (or the REST API more directly). That may still trip us up, since we use the stock kubectl library to send manifests to Kubernetes.

**Comment by wstrange** [comment]
I ran into this today with minikube vs. GKE.

I need  a storageclass object on GKE that causes issues when deployed on minikube.

**Comment by jstrachan** [comment]
@wstrange I'd expect the storageclass object to be part of the kubernetes cluster though; i.e. each cluster ideally would have dynamic PVs enabled for you; as the PV implementation is cluster specific.

I wonder if helm should assume the storageclass is setup correctly to support dynamic PVs; I raised this issue https://github.com/kubernetes/helm/issues/1583

**Comment by wstrange** [comment]
@jstrachan Yes, thinking through this some more,  the storageclass should be outside helm.

**Comment by technosophos** [comment]
The PV issue has come up several times in the Kubernetes charts repo. This one is a big pain currently because the chart user has to enable/disable PV support in a chart (if it supports that). And that is a pain. Exploring some ways to get closer to what @jstrachan asked for in terms of being able to query that information from within a chart.

**Comment by technosophos** [comment]
Okay, so we have ThirdPartyResources support basically done in 2.2.0. In 2.2.0 we also added support for any object type that the Kubernetes API definitions know about (instead of just the core ones).

Can someone verify that this is all that is needed to accomplish the goal of this ticket?

**Comment by bacongobbler** [comment]
ping @jstrachan ^^

**Comment by dhananjaysathe** [comment]
Bump has anyone got this working with openshift resources ? Would be grateful for any pointers

**Comment by RobWin** [comment]
Indeed :)

**Comment by megalord** [comment]
For what it's worth, I've been using helm to deploy resources to openshift for a few weeks now.  It isn't pretty, but you can intercept the outbound http requests from tiller to the kubernetes api by overriding the api domain using environment variables (`KUBERNETES_SERVICE_HOST` and `KUBERNETES_SERVICE_PORT`).  With that in place, it's just a matter of being able to decide which resources need to target kubernetes and which need to target openshift.  I stood up a small service to handle that logic, which I hope I can eventually get rid of if openshift finds a place in the api federation idea.

I'm happy to provide more details if desired, but I just thought I'd stop by to say this is at least possible without having to touch the helm or tiller code at all.  I did not attempt to solve for conditional application since I'm only targeting a single platform.

**Comment by gergan** [comment]
@megalord  could you provide more details on your solution?

**Comment by megalord** [comment]
Sure thing, I open sourced the code [here](https://github.com/megalord/openshift-api-group).  If anyone has further questions, just open an issue on that repo.

**Comment by gergan** [comment]
Thanks a lot @megalord!

**Comment by dhananjaysathe** [comment]
Great stuff @megalord thanks a bunch

**Comment by bacongobbler** [comment]
closing as resolved/inactive as TPR/CRD support was merged into helm 2.7 via #3019. Please re-open if this doesn't accomplish the original goal of this ticket. Thanks!
